### ISSUE TITLE
AccessViolationException in System.Net.WebSockets WebSocket.Send when client aborts

### ISSUE THREAD

**Issue by Felk** [dotnet/runtime#45934]
I am experiencing occasional `AccessViolationException`s in `System.Net.WebSockets.WebSocket`'s `SendAsync` method when the connected client is simultaneously aborting the connection. I am executing my test using .NET 5.0.100 on Windows 10 1909

The documentation states:

> Exactly one send and one receive is supported on each WebSocket object in parallel

And as far as I can tell I am not performing any concurrent reads or concurrent writes.

I was able to write a test that reproduces the issue for me, if executed often enough (every ~20th execution or so crashes, but it's very finnicky). You can try to run `while ($true) {dotnet test --filter "FullyQualifiedName=WebsocketAccessViolationTest.AccessViolationTest"}` and wait for the test to fail:

```cs
using System;
using System.Diagnostics;
using System.Net;
using System.Net.WebSockets;
using System.Text;
using System.Threading;
using System.Threading.Tasks;
using NUnit.Framework;

public class WebsocketAccessViolationTest
{
    private WebSocket? _webSocket;

    private async Task ReceiveConnections(HttpListener server)
    {
        HttpListenerContext context = await server.GetContextAsync();
        Debug.Assert(context.Request.IsWebSocketRequest);
        HttpListenerWebSocketContext webSocketContext = await context.AcceptWebSocketAsync(null!);
        WebSocket webSocket = webSocketContext.WebSocket;
        _webSocket = webSocket;
        await webSocket.ReceiveAsync(new ArraySegment<byte>(new byte[1024]), CancellationToken.None);
        Debug.Assert(false, "no messages are being received");
    }

    [Test]
    public async Task AccessViolationTest()
    {
        const int port = 12345;

        HttpListener server = new HttpListener();
        server.Prefixes.Add($"http://localhost:{port}/");
        server.Start();
        Task serverTask = ReceiveConnections(server);

        ClientWebSocket client = new ClientWebSocket();
        await client.ConnectAsync(new Uri($"ws://localhost:{port}"), CancellationToken.None);
        while (_webSocket == null)
        {
            await Task.Delay(TimeSpan.FromMilliseconds(1));
        }

        client.Abort();

        byte[] bytes = Encoding.UTF8.GetBytes("plsnocrash");
        try
        {
            // this occasionally throws
            await _webSocket!.SendAsync(bytes, WebSocketMessageType.Text, endOfMessage: true, CancellationToken.None);
        }
        catch (WebSocketException ex) when (ex.WebSocketErrorCode == WebSocketError.InvalidState)
        {
            // expected error
            // Console.Error.WriteLine(ex);
        }
    }
}
```

The full stacktrace I get is as follows:

```
WebsocketAccessViolationTest.AccessViolationTest

System.AccessViolationException : Attempted to read or write protected memory. This is often an indication that other memory is corrupt.
   at System.Net.WebSockets.WebSocketBuffer.ConvertPinnedSendPayloadToNative(Byte[] buffer, Int32 offset, Int32 count)
   at System.Net.WebSockets.WebSocketHttpListenerDuplexStream.HttpListenerAsyncEventArgs.UpdateDataChunk(Int32 index, Byte[] buffer, Int32 offset, Int32 count)
   at System.Net.WebSockets.WebSocketHttpListenerDuplexStream.HttpListenerAsyncEventArgs.UpdateDataChunk()
   at System.Net.WebSockets.WebSocketHttpListenerDuplexStream.WriteAsyncFast(HttpListenerAsyncEventArgs eventArgs)
   at System.Net.WebSockets.WebSocketHttpListenerDuplexStream.MultipleWriteAsyncCore(IList`1 sendBuffers, CancellationToken cancellationToken)
   at System.Net.WebSockets.WebSocketBase.SendFrameAsync(IList`1 sendBuffers, CancellationToken cancellationToken)
   at System.Net.WebSockets.WebSocketBase.WebSocketOperation.Process(Nullable`1 buffer, CancellationToken cancellationToken)
   at System.Net.WebSockets.WebSocketBase.SendAsyncCore(ArraySegment`1 buffer, WebSocketMessageType messageType, Boolean endOfMessage, CancellationToken cancellationToken)
   at WebsocketAccessViolationTest.AccessViolationTest() in D:\Projekte\tpp-core\Core.Tests\Overlay\WebsocketAccessViolationTest.cs:line 48
   at NUnit.Framework.Internal.TaskAwaitAdapter.GenericAdapter`1.BlockUntilCompleted()
   at NUnit.Framework.Internal.MessagePumpStrategy.NoMessagePumpStrategy.WaitForCompletion(AwaitAdapter awaitable)
   at NUnit.Framework.Internal.AsyncToSyncAdapter.Await(Func`1 invoke)
   at NUnit.Framework.Internal.Commands.TestMethodCommand.RunTestMethod(TestExecutionContext context)
   at NUnit.Framework.Internal.Commands.TestMethodCommand.Execute(TestExecutionContext context)
   at NUnit.Framework.Internal.Execution.SimpleWorkItem.PerformWork()
```

My test project is configured as follows:

```xml
<PropertyGroup>
    <TargetFramework>net5.0</TargetFramework>
    <Nullable>enable</Nullable>
    <TreatWarningsAsErrors>true</TreatWarningsAsErrors>
    <InvariantGlobalization>true</InvariantGlobalization>
</PropertyGroup>
```

**Comment by antonfirsov** [comment]
I can confirm the repro based on the code of OP.

**Comment by antonfirsov** [comment]
The `AccessViolationException` is being thrown manually, and doesn't necessarily indicate a memory corruption:

https://github.com/dotnet/runtime/blob/86224fd11c943ebe111c602358f8b8d3c3fb7494/src/libraries/System.Net.HttpListener/src/System/Net/Windows/WebSockets/WebSocketBuffer.cs#L201-L208

Looks like a corner cases scenario, and I don't know how much we want to invest into `HttpListener` which is a compat-only API so putting back to untriaged. We generally recommend to use ASP.NET for server scenarios. 

We may want to change the docs and/or add some extra validation maybe?

**Comment by karelz** [comment]
Triage: Kestrel should work fine and it does not seem to be wide impact in `HttpListener` (which is compat only API). Closing for now as Won't Fix, unless we get more reports and data about wider impact.

**Comment by Felk** [comment]
Thank you both for taking a look and spending the time to look at and evaluating this issue. I have a few follow-up questions if you don't mind me asking them in this issue, since I consider them topical here:

- Should a note be added to the documentation page for [`HttpListener`](https://docs.microsoft.com/en-us/dotnet/api/system.net.httplistener?view=net-5.0) regarding it being a "compat-only" API, and what exactly that entails? I couldn't find anything on the page that would clarify that I should have been using Kestrel instead.
- Are `HttpListener` and the accompanying websocket classes going to be marked as deprecated in a future version?
- Given my limited usecase and that Kestral is quite heavyweight compared to `HttpListener`, are there any other (officially supported) lightweight alternatives?
